walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {  
+    long meeting_duration = request.getDuration();
+    Collection<String> meeting_attendees = request.getAttendees();
+
+    //First, find all incompatible time ranges so that we can use that as reference
+    //for finding time ranges which do actually work.
+    ArrayList<TimeRange> incompatible_ranges = findIncompatibleRanges(events, meeting_attendees);
+
+    //Sort the incompatible ranges by start time.
+    Collections.sort(incompatible_ranges, TimeRange.ORDER_BY_START);
+    Collection<TimeRange> available_ranges = new ArrayList<TimeRange>();
+
+    //Check if there is enough time between the end of the last event and the beginning of the next
+    //event. This is initalized to 0 since we are checking at the start of the day.
+    int meeting_request_start_time = 0; 
+    for (TimeRange event_range : incompatible_ranges){
+      int range_end = event_range.end();
+      int range_start = event_range.start();
+        if (!(meeting_request_start_time + meeting_duration > range_start)){ 
+          TimeRange valid_range = TimeRange.fromStartEnd(meeting_request_start_time, range_start, false);
+          available_ranges.add(valid_range);
+        }
+      
+      //Handles cases where events overlap and the latest ending event time is actually less than the
+      //previous event end time. 
+      meeting_request_start_time = Math.max(range_end,meeting_request_start_time);       
+    }
+
+    //the potential meeting time of the last possible slot will be whatever time the last event ended up until
+    //midnight. 
+    if (meeting_request_start_time + meeting_duration <= 24*60){
```

Comment:
Numbers like this are usually stored in a variable at the top to make their meaning more explicit. Something like:

```
int MINUTES_IN_DAY = 24 * 60;
```